File: iOS_demo/UyaliBeautyFaceSDKDemo-OC/UyaliBeautyFaceSDKDemo-OC/美颜/FaceReshapeView.m
//
//  FaceReshapeView.m
//  UyaliBeautyFaceSDKDemo-OC
//
//  Created by S weet on 2023/2/27.
//

#import "FaceReshapeView.h"
#import "UyaliButton.h"

@interface FaceReshapeView ()

@property (nonatomic, strong)NSArray *beautyItems;

@property (nonatomic, assign) NSInteger currentTag;

@end

@implementation FaceReshapeView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self initDatas];
        [self setupView];
    }
    return self;
}

- (void)initDatas {
    
    _beautyItems = @[
        @{@"image": @"beauty_shape_head_reduce", @"name": @"小头"},
        @{@"image": @"beauty_shape_face_thin", @"name": @"瘦脸"},
        @{@"image": @"beauty_shape_face_narrow", @"name": @"窄脸"},
        @{@"image": @"beauty_shape_face_v", @"name": @"V脸"},
        @{@"image": @"beauty_shape_face_small", @"name": @"小脸"},
        @{@"image": @"beauty_shape_chin", @"name": @"下巴"},
        @{@"image": @"beauty_shape_forehead", @"name": @"额头"},
        @{@"image": @"beauty_shape_cheekbone", @"name": @"颧骨"},
        @{@"image": @"beauty_shape_eye_big", @"name": @"大眼"},
        @{@"image": @"beauty_shape_eye_distance", @"name": @"眼距"},
        @{@"image": @"beauty_shape_eye_corner", @"name": @"开眼角"},
        @{@"image": @"beauty_shape_eyelid_down", @"name": @"眼睑下至"},
        @{@"image": @"beauty_shape_nose_thin", @"name": @"瘦鼻"},
        @{@"image": @"beauty_shape_nose_wing", @"name": @"鼻翼"},
        @{@"image": @"beauty_shape_nose_long", @"name": @"长鼻"},
        @{@"image": @"beauty_shape_nose_root", @"name": @"山根"},
        @{@"image": @"beauty_shape_eyebrow_distance", @"name": @"眉间距"},
        @{@"image": @"beauty_shape_eyebrow_thin", @"name": @"眉粗细"},
        @{@"image": @"beauty_shape_mouth", @"name": @"嘴型"}];
    
    _currentTag = 100;
}

#pragma mark - Setter & Getter
- (void)setCurrentTag:(NSInteger)currentTag {
    if (_currentTag != currentTag) {
        UyaliButton *newBtn = (UyaliButton *)[self viewWithTag:currentTag];
        UyaliButton *oldBtn = (UyaliButton *)[self viewWithTag:_currentTag];
        
        newBtn.selected = YES;
        oldBtn.selected = NO;
        
        if (currentTag == 100) {//小头 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 101) {//瘦脸 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 102) {//窄脸 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 103) {//V脸 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 104) {//小脸 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 105) {//下巴 done -50 - 50
            _slider.minimumValue = -50;
            _slider.maximumValue = 50;
        } else if (currentTag == 106) {//额头 done -50 - 50
            _slider.minimumValue = -50;
            _slider.maximumValue = 50;
        } else if (currentTag == 107) {//颧骨 done -50 - 50
            _slider.minimumValue = -50;
            _slider.maximumValue = 50;
        } else if (currentTag == 108) {//大眼 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 109) {//眼距 done -50 - 50
            _slider.minimumValue = -50;
            _slider.maximumValue = 50;
        } else if (currentTag == 110) {//眼角 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 111) {//下眼睑 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 112) {//瘦鼻 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 113) {//鼻翼 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 114) {//长鼻 done -50 - 50
            _slider.minimumValue = -50;
            _slider.maximumValue = 50;
        } else if (currentTag == 115) {//鼻子山根 done 0 - 100
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        } else if (currentTag == 116) {//眉间距 done -50 - 50
            _slider.minimumValue = -50;
            _slider.maximumValue = 50;
        } else if (currentTag == 117) {//眉粗细 done -50 - 50
            _slider.minimumValue = -50;
            _slider.maximumValue = 50;
        } else if (currentTag == 118) {//嘴型 done -50 - 50
            _slider.minimumValue = -50;
            _slider.maximumValue = 50;
        }
        
        _currentTag = currentTag;
    }
}

#pragma mark - UI
- (void)setupView {
    self.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.4];
    
    _slider = [[UISlider alloc]initWithFrame:CGRectMake(16, 24, self.bounds.size.width-32, 30)];
    [self addSubview:_slider];
    [_slider addTarget:self action:@selector(sliderValueChanged:) forControlEvents:UIControlEventValueChanged];
    
    _valueLabel = [[UILabel alloc]initWithFrame:CGRectMake(0, 0, self.bounds.size.width, _slider.frame.origin.y)];
    [self addSubview:_valueLabel];
    _valueLabel.font = [UIFont systemFontOfSize:15 weight:UIFontWeightMedium];
    _valueLabel.textColor = [UIColor whiteColor];
    _valueLabel.textAlignment = NSTextAlignmentCenter;
    _valueLabel.text = @"0.0";
    
    UIScrollView *scroll = [[UIScrollView alloc]initWithFrame:CGRectMake(0, 60, self.bounds.size.width, self.bounds.size.height-70)];
    [self addSubview:scroll];
    
    CGFloat buttonWidth = scroll.bounds.size.height;
    CGFloat buttonHeight = scroll.bounds.size.height;
    CGFloat buttonMargin = 16;
    
    scroll.contentSize = CGSizeMake(buttonMargin+(buttonWidth+buttonMargin)*_beautyItems.count, scroll.frame.size.height);
    
    for (NSInteger i = 0; i < _beautyItems.count; i++) {
        NSDictionary *item = _beautyItems[i];
        UyaliButton *button = [[UyaliButton alloc]initWithFrame:CGRectMake(buttonMargin+(buttonMargin+buttonWidth)*i, 0, buttonWidth, buttonHeight)];
        [scroll addSubview:button];
        [button setTitle:item[@"name"] forState:UIControlStateNormal];
        [button setImage:[UIImage imageNamed:item[@"image"]] forState:UIControlStateNormal];
        [button addTarget:self action:@selector(reshapeButtonAction:) forControlEvents:UIControlEventTouchUpInside];
        button.tag = 100+i;
        if (i == 0) {
            button.selected = YES;
            
            _slider.minimumValue = 0;
            _slider.maximumValue = 100;
        }
        button.layer.cornerRadius = buttonWidth/2;
    }
}

#pragma mark - Action
- (void)reshapeButtonAction:(UyaliButton *)button {
    self.currentTag = button.tag;
    if (self.delegate != nil && [self.delegate respondsToSelector:@selector(changeReshapeType:)]) {
        [self.delegate changeReshapeType:button.tag];
    }
}

- (void)sliderValueChanged:(UISlider *)currentSlider {
    _valueLabel.text = [NSString stringWithFormat:@"%.1f",currentSlider.value];
    if (self.delegate != nil && [self.delegate respondsToSelector:@selector(getReshapeDeltaVale:type:)]) {
        [self.delegate getReshapeDeltaVale:_slider.value type:_currentTag];
    }
}

@end

File: iOS_demo/UyaliBeautyFaceSDKDemo/UyaliBeautyFaceSDKDemo/FaceReshapeView.swift
//
//  FaceReshapeView.swift
//  UyaliBeautyFaceSDKDemo
//
//  Created by S weet on 2023/1/28.
//

import UIKit

@objc protocol FaceReshapeDelegate: NSObjectProtocol {
    func changeReshapeType(type: Int)
    func getDeltaValue(value:Float, type:Int)
}

class FaceReshapeView: UIView {
    
    weak var delegate: FaceReshapeDelegate!
    
    var slider: UISlider!
    var valueLabel: UILabel!
    
    private let reshapeItems: [[String: String]] = [
        ["image": "beauty_shape_head_reduce", "name": "小头"],
        ["image": "beauty_shape_face_thin", "name": "瘦脸"],
        ["image": "beauty_shape_face_narrow", "name": "窄脸"],
        ["image": "beauty_shape_face_v", "name": "V脸"],
        ["image": "beauty_shape_face_small", "name": "小脸"],
        ["image": "beauty_shape_chin", "name": "下巴"],
        ["image": "beauty_shape_forehead", "name": "额头"],
        ["image": "beauty_shape_cheekbone", "name": "颧骨"],
        ["image": "beauty_shape_eye_big", "name": "大眼"],
        ["image": "beauty_shape_eye_distance", "name": "眼距"],
        ["image": "beauty_shape_eye_corner", "name": "开眼角"],
        ["image": "beauty_shape_eyelid_down", "name": "眼睑下至"],
        ["image": "beauty_shape_nose_thin", "name": "瘦鼻"],
        ["image": "beauty_shape_nose_wing", "name": "鼻翼"],
        ["image": "beauty_shape_nose_long", "name": "长鼻"],
        ["image": "beauty_shape_nose_root", "name": "山根"],
        ["image": "beauty_shape_eyebrow_distance", "name": "眉间距"],
        ["image": "beauty_shape_eyebrow_thin", "name": "眉粗细"],
        ["image": "beauty_shape_mouth", "name": "嘴型"]
    ]
    
    private var currentTag: Int = 100 {
        didSet {
            if currentTag != oldValue {
                let newBtn = viewWithTag(currentTag) as! UyaliButton
                let oldBtn = viewWithTag(oldValue) as! UyaliButton
                
                newBtn.isSelected = true
                oldBtn.isSelected = false
                
                if currentTag == 100 {//小头 done 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                } else if currentTag == 101 {//瘦脸 done 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                } else if currentTag == 102 {//窄脸 done 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                } else if currentTag == 103 {//V脸 done 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                } else if currentTag == 104 {//小脸 done 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                } else if currentTag == 105 {//下巴 done -50 - 50
                    slider.minimumValue = -50
                    slider.maximumValue = 50
                } else if currentTag == 106 {//额头 done -50 - 50
                    slider.minimumValue = -50
                    slider.maximumValue = 50
                } else if currentTag == 107 {//颧骨 done -50 - 50
                    slider.minimumValue = -50
                    slider.maximumValue = 50
                } else if currentTag == 108 {//大眼 done 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                } else if currentTag == 109 {//眼距 done -50 - 50
                    slider.minimumValue = -50
                    slider.maximumValue = 50
                } else if currentTag == 110 {//眼角 done 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                } else if currentTag == 111 {//下眼睑 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                } else if currentTag == 112 {//瘦鼻 done 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                } else if currentTag == 113 {//鼻翼 done 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                } else if currentTag == 114 {//长鼻 done -50 - 50
                    slider.minimumValue = -50
                    slider.maximumValue = 50
                } else if currentTag == 115 {//鼻子山根 done 0 - 100
                    slider.minimumValue = 0
                    slider.maximumValue = 100
                } else if currentTag == 116 {//眉间距 done -50 - 50
                    slider.minimumValue = -50
                    slider.maximumValue = 50
                } else if currentTag == 117 {//眉粗细 done -50 - 50
                    slider.minimumValue = -50
                    slider.maximumValue = 50
                } else if currentTag == 118 {//嘴型 done -50 - 50
                    slider.minimumValue = -50
                    slider.maximumValue = 50
                }
            }
        }
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
        setupView()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    //MARK: UI
    func setupView() {
        self.backgroundColor = .black.withAlphaComponent(0.4)
        
        slider = UISlider(frame: CGRect(x: 16, y: 24, width: bounds.width-32, height: 30))
        addSubview(slider)
        slider.addTarget(self, action: #selector(sliderValueChanged(currentSlider: )), for: .valueChanged)
        
        valueLabel = UILabel(frame: CGRect(x: 0, y: 0, width: bounds.width, height: slider.frame.origin.y))
        addSubview(valueLabel)
        valueLabel.font = .systemFont(ofSize: 15, weight: .medium)
        valueLabel.textColor = .white
        valueLabel.textAlignment = .center
        valueLabel.text = "0.0"
        
        let scroll = UIScrollView(frame: CGRect(x: 0, y: 60, width: bounds.width, height: bounds.height-70))
        addSubview(scroll)
        
        let buttonWidth = scroll.bounds.height
        let buttonHeight = scroll.bounds.height
        let buttonMargin = 16.0
        
        scroll.contentSize = CGSize(width: buttonMargin+(buttonWidth+buttonMargin)*CGFloat(reshapeItems.count), height: scroll.frame.height)
        
        for i in 0..<reshapeItems.count {
            let item = reshapeItems[i]
            let button = UyaliButton(frame: CGRect(x: buttonMargin+(buttonMargin+buttonWidth)*CGFloat(i), y: 0, width: buttonWidth, height: buttonHeight))
            scroll.addSubview(button)
            button.setTitle(item["name"], for: .normal)
            button.setImage(UIImage(named: item["image"]!), for: .normal)
            button.addTarget(self, action: #selector(reshapeButtonAction(button:)), for: .touchUpInside)
            button.tag = 100+i
            if i == 0 {
                button.isSelected = true
                currentTag = 100
                
                slider.minimumValue = 0
                slider.maximumValue = 100
            }
            button.layer.cornerRadius = buttonWidth/2
        }
    }
    
    //MARK: Action
    @objc func reshapeButtonAction(button: UyaliButton) {
        currentTag = button.tag
        if delegate != nil {
            delegate.changeReshapeType(type: button.tag)
        }
    }
    
    @objc func sliderValueChanged(currentSlider: UISlider) {
        valueLabel.text = String(format: "%.1f", currentSlider.value)
        if delegate != nil {
            delegate.getDeltaValue(value: slider.value, type: currentTag)
        }
    }
}


class UyaliButton: UIButton {
    override init(frame: CGRect) {
        super.init(frame: frame)
            
        setupView()
    }
        
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override var isSelected: Bool {
        didSet {
            if isSelected {
                self.backgroundColor = .black.withAlphaComponent(0.4)
                self.imageView?.tintColor = .systemBlue
            } else {
                self.backgroundColor = .white.withAlphaComponent(0.4)
                self.imageView?.tintColor = .white
            }
        }
    }
        
    func setupView() {
        self.backgroundColor = .white.withAlphaComponent(0.4)
        self.titleLabel?.textAlignment = .center
        self.imageView?.contentMode = .scaleAspectFit
        self.titleLabel?.font = UIFont.systemFont(ofSize: 12)
        self.imageView?.tintColor = .white
        
        self.setTitleColor(.white, for: .normal)
        self.setTitleColor(.systemBlue, for: .selected)
    }
        
    override func titleRect(forContentRect contentRect: CGRect) -> CGRect {
        let titleX = 0
        let titleY = contentRect.size.height * 0.5
        let titleW = contentRect.size.width
        let titleH = contentRect.size.height - titleY
        return CGRect(x: CGFloat(titleX), y: titleY, width: titleW, height: titleH)
    }
        
    override func imageRect(forContentRect contentRect: CGRect) -> CGRect {
        let imageW = contentRect.width
        let imageH = contentRect.size.height * 0.5
        return CGRect(x: 0, y: 5, width: imageW, height: imageH)
    }
}
